### PR TITLE
Support Custom Split Config for Trainer and Tuner Components

### PR DESCRIPTION
**Pull request by tinally**
# Support custom split config for Trainer and Tuner components by adding splits param to TrainArgs and EvalArgs.

### CODE REVIEW

File: tfx/components/trainer/fn_args_utils.py [hunk 1]
```diff
@@ -86,6 +75,24 @@ def get_common_fn_args(input_dict: Dict[Text, List[types.Artifact]],
   json_format.Parse(exec_properties[constants.TRAIN_ARGS_KEY], train_args)
   json_format.Parse(exec_properties[constants.EVAL_ARGS_KEY], eval_args)
 
+  # Default behavior (when train_args.splits and eval_args.splits are not set)
+  # is train on ‘train’ splits and evaluate on ‘eval’ splits.
+  if not train_args.splits:
+    train_args.splits = 'train'
```
Comment (1025KB) [inline tfx/components/trainer/fn_args_utils.py:81]: is splits a list of str?

keep existing test cases (covers default split code path), and create a new test covers custom split path

File: tfx/proto/trainer.proto [hunk 2]
```diff
@@ -17,16 +17,22 @@ package tfx.components.trainer;
 
 // Args specific to training.
 message TrainArgs {
+  // The example split names to train on.
+  string splits = 1;
```
Comment (1025KB) [inline tfx/proto/trainer.proto:21]: it's repeated string (list of splits), same for eval

Comment (tinally) [reply tfx/proto/trainer.proto]: got it, i changed it, thanks for the catch

File: tfx/components/trainer/component.py [hunk 3]
```diff
@@ -198,8 +200,8 @@ def trainer_fn(trainer.executor.TrainerFnArgs,
     """
     if [bool(module_file), bool(run_fn), bool(trainer_fn)].count(True) != 1:
       raise ValueError(
-          "Exactly one of 'module_file', 'trainer_fn', or 'run_fn' must be supplied."
-      )
+          "Exactly one of 'module_file', 'trainer_fn', or 'run_fn' must be "
+          "supplied.")
```
Comment (1025KB) [inline tfx/components/trainer/component.py:204]: why change this?

File: tfx/components/trainer/executor.py [hunk 4]
```diff
@@ -178,8 +178,8 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
 
     Args:
       input_dict: Input dict from input key to a list of ML-Metadata Artifacts.
-        - examples: Examples used for training, must include 'train' and 'eval'
-          splits.
+        - examples: Examples used for training, support custom splits. If not
+           specified, include 'train' and 'eval' splits by default.
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:182]: how about this[1] to be more specific.

[1] examples: Examples used for training, must include 'train' and 'eval' if custom splits is not specified in train_args and eval_args.

File: tfx/components/trainer/fn_args_utils.py [hunk 5]
```diff
@@ -86,6 +75,27 @@ def get_common_fn_args(input_dict: Dict[Text, List[types.Artifact]],
   json_format.Parse(exec_properties[constants.TRAIN_ARGS_KEY], train_args)
   json_format.Parse(exec_properties[constants.EVAL_ARGS_KEY], eval_args)
 
+  # Default behavior is train on ‘train’ splits (when splits is empty in train
+  # args) and evaluate on 'eval' splits (when splits is empty in eval args).
+  if not train_args.splits:
+    train_args.splits.append('train')
```
Comment (1025KB) [inline tfx/components/trainer/fn_args_utils.py:81]: just confirm, when it's not set, train_args.splits is [] not None, right?

File: tfx/components/tuner/component_test.py [hunk 6]
```diff
@@ -81,6 +81,18 @@ def testConstructDuplicateUserModule(self):
           module_file='/path/to/module/file',
           tuner_fn='path.to.tuner_fn')
 
+  def testConstructWithCustomSplits(self):
+    self.train_args = trainer_pb2.TrainArgs(splits=['train'], num_steps=100)
+    self.eval_args = trainer_pb2.EvalArgs(splits=['eval'], num_steps=50)
```
Comment (1025KB) [inline tfx/components/tuner/component_test.py:86]: you can just update line37&38 and drop this test, same for the Trainer

this is because we didn't do any special handling in component.py

File: tfx/components/tuner/executor_test.py [hunk 7]
```diff
@@ -127,6 +127,24 @@ def testTuneArgs(self):
...
+    self._exec_properties['module_file'] = os.path.join(self._testdata_dir,
+                                                        'module_file',
+                                                        'tuner_module.py')
+
+    tuner = executor.Executor(self._context)
+    tuner.Do(
+        input_dict=self._input_dict,
+        output_dict=self._output_dict,
+        exec_properties=self._exec_properties)
+
```
Comment (1025KB) [inline tfx/components/tuner/executor_test.py:147]: no assert?

File: tfx/components/trainer/component.py (hunk 3 above)
Comment (tinally) [reply tfx/components/trainer/component.py:204]: this line throws a pylint error in the previous commits

File: tfx/components/trainer/fn_args_utils.py (hunk 5 above)
Comment (tinally) [reply tfx/components/trainer/fn_args_utils.py:81]: yes, because splits is a repeated string, so the default is [] when splits is not provided by user

File: tfx/components/trainer/executor.py (hunk 4 above)
Comment (tinally) [reply tfx/components/trainer/executor.py:182]: yes, i think this looks more clear, i will update this

File: tfx/components/tuner/component.py [hunk 8]
```diff
@@ -84,9 +84,11 @@ def tuner_fn(fn_args: FnArgs) -> TunerFnResult:
         'module_file' for the required signature of the UDF. Exactly one of
         'module_file' or 'tuner_fn' must be supplied.
       train_args: A trainer_pb2.TrainArgs instance, containing args used for
-        training. Current only num_steps is available.
+        training. Current only splits and num_steps are available. Default
+        behavior (when splits is empty) is train on 'train' splits.
```
Comment (goutham) [inline tfx/components/tuner/component.py:88]: ... is train on 'train' split.

("split" singular)

File: tfx/components/tuner/component.py [hunk 9]
```diff
@@ -84,9 +84,11 @@ def tuner_fn(fn_args: FnArgs) -> TunerFnResult:
...
       eval_args: A trainer_pb2.EvalArgs instance, containing args used for eval.
-        Current only num_steps is available.
+        Current only splits and num_steps are available. Default behavior (when
+        splits is empty) is evaluate on ‘eval’ splits.
```
Comment (goutham) [inline tfx/components/tuner/component.py:91]: ... is evaluate on 'eval' split.

("split" singular)

File: tfx/components/trainer/executor_test.py [hunk 10]
```diff
@@ -186,6 +186,19 @@ def testDoWithHyperParameters(self):
     self._verify_model_exports()
     self._verify_model_run_exports()
 
+  def testDoWithCustomSplits(self):
```
Comment (goutham) [inline tfx/components/trainer/executor_test.py:189]: Test is for custom splits but it's just using the default 'train' and 'eval' splits. So test results would be no different if those were omitted. Better to use actual custom splits.

File: tfx/components/tuner/executor_test.py [hunk 11]
```diff
@@ -127,6 +127,26 @@ def testTuneArgs(self):
           output_dict=self._output_dict,
           exec_properties=self._exec_properties)
 
+  def testDoWithCustomSplits(self):
```
Comment (goutham) [inline tfx/components/tuner/executor_test.py:130]: Shouldn't this test with actual custom splits (different from "train" and "eval")? Otherwise test results would be no different if the code actually ignores "splits" entirely.

File: tfx/components/trainer/component.py [hunk 12]
```diff
@@ -169,13 +169,15 @@ def trainer_fn(trainer.executor.TrainerFnArgs,
         based trainer. See 'module_file' for the required signature of the UDF.
         Exactly one of 'module_file' or 'trainer_fn' must be supplied.
       train_args: A trainer_pb2.TrainArgs instance or a dict, containing args
-        used for training. Current only num_steps is available. If it's provided
-        as a dict and any field is a RuntimeParameter, it should have the same
-        field names as a TrainArgs proto message.
+        used for training. Current only splits and num_steps are available. If
+        it's provided as a dict and any field is a RuntimeParameter, it should
+        have the same field names as a TrainArgs proto message. Default
+        behavior (when splits is empty) is train on ‘train’ splits.
```
Comment (goutham) [inline tfx/components/trainer/component.py:175]: ... is train on 'train' split.

("split" singular)

File: tfx/components/trainer/component.py [hunk 13]
```diff
@@ -169,13 +169,15 @@ def trainer_fn(trainer.executor.TrainerFnArgs,
...
       eval_args: A trainer_pb2.EvalArgs instance or a dict, containing args
-        used for evaluation. Current only num_steps is available. If it's
-        provided as a dict and any field is a RuntimeParameter, it should have
-        the same field names as a EvalArgs proto message.
+        used for evaluation. Current only splits and num_steps are available.
+        If it's provided as a dict and any field is a RuntimeParameter, it
+        should have the same field names as a EvalArgs proto message. Default
+        behavior (when splits is empty) is evaluate on ‘eval’ splits.
```
Comment (goutham) [inline tfx/components/trainer/component.py:180]: ... is evaluate on 'eval' split.

("split" singular)

File: tfx/components/trainer/component.py [hunk 14]
```diff
@@ -169,13 +169,15 @@ def trainer_fn(trainer.executor.TrainerFnArgs,
...
-        field names as a TrainArgs proto message.
+        used for training. Current only splits and num_steps are available. If
+        it's provided as a dict and any field is a RuntimeParameter, it should
+        have the same field names as a TrainArgs proto message. Default
+        behavior (when splits is empty) is train on ‘train’ splits.
       eval_args: A trainer_pb2.EvalArgs instance or a dict, containing args
-        used for evaluation. Current only num_steps is available. If it's
-        provided as a dict and any field is a RuntimeParameter, it should have
-        the same field names as a EvalArgs proto message.
+        used for evaluation. Current only splits and num_steps are available.
```
Comment (goutham) [inline tfx/components/trainer/component.py:177]: Current -> Currently

File: tfx/components/trainer/component.py [hunk 15]
```diff
@@ -169,13 +169,15 @@ def trainer_fn(trainer.executor.TrainerFnArgs,
         based trainer. See 'module_file' for the required signature of the UDF.
         Exactly one of 'module_file' or 'trainer_fn' must be supplied.
       train_args: A trainer_pb2.TrainArgs instance or a dict, containing args
-        used for training. Current only num_steps is available. If it's provided
-        as a dict and any field is a RuntimeParameter, it should have the same
-        field names as a TrainArgs proto message.
+        used for training. Current only splits and num_steps are available. If
```
Comment (goutham) [inline tfx/components/trainer/component.py:172]: Current -> Currently

File: tfx/components/tuner/component.py (hunk 8 above)
Comment (tinally) [reply tfx/components/tuner/component.py:88]: got it, thanks for the catch, so detailed!

File: tfx/components/trainer/executor_test.py (hunk 10 above)
Comment (tinally) [reply tfx/components/trainer/executor_test.py:189]: To use custom splits, we need to add a custom split folder to the iris data. Currently, the tfx/components/testdata/iris/data folder only contains 'eval' and 'train'. Should I add another folder under testdata for this test?

Comment (goutham) [reply tfx/components/trainer/executor_test.py:189]: @1025KB Do you have a preference?

Either options sound good to me.

File: tfx/components/tuner/executor_test.py (hunk 11 above)
Comment (tinally) [reply tfx/components/tuner/executor_test.py:131]: I updated the tests for custom splits. Can you please check if it's good now?

File: tfx/components/trainer/component_test.py [hunk 16]
```diff
@@ -69,8 +69,8 @@ def testConstructWithParameter(self):
         transformed_examples=self.examples,
         transform_graph=self.transform_output,
         schema=self.schema,
-        train_args=dict(num_steps=n_steps),
-        eval_args=dict(num_steps=n_steps))
+        train_args=dict(splits='train', num_steps=n_steps),
```
[the remark this answers is not in the document]
Comment (tinally) [reply tfx/components/trainer/component_test.py]: i think i updated this last week, can you refresh the code?

File: tfx/components/tuner/executor_test.py (hunk 11 above)
Comment (goutham) [reply tfx/components/tuner/executor_test.py:130]: It's still not clear to me how we test that the custom splits are being used. If tuner executor ignored 'training' and 'evaluating' splits and instead used 'train' and 'eval' splits, would this test fail?

Comment (1025KB) [reply tfx/components/tuner/executor_test.py:130]: basically the split here should match the input example artifact, the test example only contains `train` and `eval` now.

This function is more like testDoWithSpecifiedSplits to make sure the user can specify the split by config (thus custom split names can work if examplegen generates custom splits).

Comment (goutham) [reply tfx/components/tuner/executor_test.py:130]: Let me rephrase my question: If the code being tested completely ignored TrainArgs.splits field, so instead of "training" and "evaluating" splits it used the default "train" and "eval" splits, will this test fail?

Comment (1025KB) [reply tfx/components/tuner/executor_test.py:130]: line138 changed the split name of example artifact, if train and eval is used, it will fail.

Tina, we can also remove train and eval data to make sure

Comment (goutham) [reply tfx/components/tuner/executor_test.py:130]: thanks for confirming

Comment (tinally) [reply tfx/components/tuner/executor_test.py:130]: it's updated now, i created a new example artifact so in the data folder there's only 'training' and 'evaluating', if it used 'train' and 'eval', it will fail.
